### ISSUE TITLE
[Bug] Strange FIFOQueue behavior in distributed tensorflow. Is this a bug?

### ISSUE THREAD

**Issue by DevarakondaV** [tensorflow/tensorflow#22878]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: pip install tensorflow 
- **TensorFlow version (use command below)**: 1.8
- **Python version**: 3.6.5
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: No CUDA
- **GPU model and memory**: No GPU
- **Exact command to reproduce**: Check code below

------------------------
### Describe the problem
There seems to be very little documentation on distributed tensorflow, so I am unsure of whether or not this is a bug or a logical error on my part. But the behavior is still a little strange.  Essentially, the issue seems to arise due to tf.train.SummarySaverHook(), it's interaction with tf.train.MoniteredTrainingSession() and the chief worker. If you look at my code below, you will see that there are two workers in this model. The chief worker is associated with "task:0". I've set up the model such that the queue and it operations live on "task:1"(not the chief worker). There is a simple sum operation on "task:0". Once I launch the 3 nodes(ps, task0, task1), task0 is able to complete one simple sum operation and then all tasks hang indefinitely.  

For the moment, I've bypassed this issue by passing the summary hook object to the non-chief worker(task:0). By doing this, there don't seem to be any issues and the code works fine. I'm unsure of why this bypasses the issue. If chief workers handle summary write operations, shouldn't it still hang?

The strange behavior that I was talking about earlier occurs when you pass both the summary hook to the chief worker (Doing this will cause all workers to hang). You can get past the issue by simple removing the tf.FIFOqueue().dequeue() operation. It seems that this operation is getting called by "task:0" despite the fact that I've specified its location as "task:1" using tf.device. It is also being called despite not being an input requirement for any part of the graph that lives on "task:0". 

My question is essentially: Is this a bug or a logical error on my part?

### Source code / logs

```
import tensorflow as tf 
import sys

s_name = str(sys.argv[1])
t_num = int(sys.argv[2])


#Cluster details
cl_spec = tf.train.ClusterSpec({
    "worker": [
        "localhost:2223",
        "localhost:2224"
    ],
    "ps": [
        "localhost:2222"
    ]
})

config = tf.ConfigProto()
config.gpu_options.allow_growth = True


#Decides whether parameter server or worker
if s_name == "ps":
    server = tf.train.Server(cl_spec,job_name="ps",task_index=0,config=config)
    server.join()
else:
    server = tf.train.Server(cl_spec,job_name="worker",task_index=t_num,config=config)

    with tf.device("/job:worker/replica:0/task:0"):
        a = tf.Variable(tf.constant(5))
        b = tf.Variable(tf.constant(6))
        z = tf.add(a,b)
        z_p = tf.Print([z],[z],"sum: ")
        tf.summary.histogram("t0sum",z_p)

    with tf.device("/job:ps/replica:0/task:0"):    
        with tf.name_scope("train_place_holder"):
            x = tf.placeholder(tf.uint8,shape=[1,1],name="s_img1")
            y = tf.placeholder(tf.uint8,shape=[1,1],name="s_a")    
        
        global_step = tf.train.create_global_step()
    
    with tf.device("/job:worker/replica:0/task:1"):
        q = tf.FIFOQueue(capacity=25,
                         dtypes= (tf.uint8,tf.uint8),
                         shapes= (tf.TensorShape([1,1]),tf.TensorShape([1,1])),
                         name="tq",shared_name="train_queue")
        
        enqueue_op = q.enqueue((x,y),name="enqueue")
        xx,yy = q.dequeue(name="dq")
        zz = tf.add(xx,yy)
        tf.summary.histogram("t1sum: ",zz)

    summ = tf.summary.merge_all()

    lap_dir = r'C:\Users\Vishnu\Documents\EngProj\bug_tf\log'

    summary_hook = tf.train.SummarySaverHook(   save_steps=1,save_secs=None,
                                                    output_dir=lap_dir,summary_writer=None,
                                                    scaffold=None,summary_op=summ)

    if (t_num == 0):
        # THis is chief worker
        saver_hook = tf.train.CheckpointSaverHook(  checkpoint_dir=lap_dir,
                                                    save_secs=3600,save_steps=None,
                                                    saver=tf.train.Saver(),checkpoint_basename='model.ckpt',
                                                    scaffold=None)

        with tf.train.MonitoredTrainingSession(master=server.target,is_chief=True,
                                                hooks=[saver_hook,summary_hook], config=config) as sess:   
           while True:
               sess.run([z,z_p])

    else:

        with tf.train.MonitoredTrainingSession(master=server.target,is_chief=False,
                                                save_summaries_steps=1,config=config) as sess:
            while not sess.should_stop():
                sess.run([zz])

    
```

You can run the above file and start 3 workers using the following command line commands.

```
python [file_name] worker 1

python [file_name] worker 0

python [file_name] ps 0
```

**Comment by DevarakondaV** [comment]
I've also noticed that adding `checkpoint_dir` argument to the chief worker session also causes the workers to hang.

**Comment by wt-huang** [comment]
@DevarakondaV Have you observed this FIFOQueue behavior when GPU is installed?

**Comment by DevarakondaV** [comment]
@wt-huang I am not using a GPU and have not tested with one.

**Comment by DevarakondaV** [comment]
Woops. I think i accidently closed the issue. Sorry.

**Comment by wt-huang** [comment]
@DevarakondaV It shows in your code snippet that you are using GPU, probably double check here:
```
config.gpu_options.allow_growth = True
```
Check the communications between ps and workers as well as how you set up the queue. You can always monitor the ps and worker performance online.

Also, make sure the chief worker is consistent in the setup, notice some discrepancy here:

> The chief worker is associated with "task:0". I've set up the model such that the queue and it operations live on "task:1"(not the chief worker).

> For the moment, I've bypassed this issue by passing the summary hook object to the non-chief worker(task:0).

**Comment by DevarakondaV** [comment]
@wt-huang

`
config.gpu_options.allow_growth = True
`
I've configured this option but I know for a fact that I am not using a GPU. 
I've also removed it before running the code again and it still hangs. Have you been able to replicate the issue? 

I should note that I am still running this distributed architecture completely on a local machine. Would this have an effect on whether or not the program hangs? (I am still a beginner at distributed computing in general).

> Also, make sure the chief worker is consistent in the setup, notice some discrepancy here:

That's totally my fault. Sorry for the confusion this may have caused. But this sentence

>For the moment, I've bypassed this issue by passing the summary hook object to the non-chief worker(task:0).

Should really say

>For the moment, I've bypassed this issue by passing the summary hook object to the non-chief worker(task:1).

I've edited the original description to reflect the change.
